Simplify switching the astro.io backend

hamogu opened the pull request:
# Summary

Add dummy_module, docs, and test case for switching the `sherpa.astro.io` backend. io backend are now tried in a default order (crates, pyfits, dummy), so this effects users who have both crates and astropy installed in a CIAO installation and manually selected the pyfits backend in their `sherpa.rc` - they will now get crates as I/O backend.

# Details
Switching this has always been possible, but this PR makes it simplerby (a) adding a dummy_backend (similar to what we already have for plotting) to ensure that sherpa.astro.io can always be imported even if no fits reader  is present, and (b) adding docs and tests on how to switch.

hamogu commented:
Note that because of https://github.com/sherpa/sherpa/issues/348#issuecomment-858630849 this is limited to dealing with the IO backend and I'll tackle the plotting backend in a different PR. (I admit that this PR is a by-product and I really want the plotting backend to be easily changable to help with development on #499).

DougBurke commented on MANIFEST.in:
```diff
@@ -36,7 +36,6 @@ include sherpa/optmethods/tests/tstopt.hh
 
 # Sherpa configuration files
 include sherpa/sherpa.rc
-include sherpa/sherpa-standalone.rc
```
I think I'd rather see this without removing `sherpa-standalone.rc` as it's a separate issue which I think would just slow this PR down,

DougBurke commented on sherpa/astro/io/tests/test_backend_switch.py:
```diff
@@ -0,0 +1,50 @@
+
+    with pytest.raises(NotImplementedError,
+                       match="No usable I/O backend was imported."):
+        pha = make_data_path("3c273.pi")
```
We don't need to re-generate `pha` here (ie this line and line 49) can be removed.

hamogu replied:
Yes, we do. The point here is to test that reading works when a fits backend is present (`@requires_fits`, reading in line 35), then we switch to the dummy backend, so we check in this line that reading now raises an error (that way, we make sure that the backend was actually switched and we don't accidentally continue to use crates or pyfits (which would not issue this error) and then we switch back to the previous backend and check in L 49 that the switch back worked.

We are not reading the pha because we want to do anything, we read it three times to make sure we can or cannot read it with the backend selected at the time.

hamogu replied:
Oh, now I understand; Im just a little dumb this weekend. Please ignore previous comment. Will fix.

hamogu replied:
Oh, now I understand. Please ignore previous comment. Will fix.

DougBurke commented on sherpa/astro/io/tests/test_backend_switch.py:
```diff
@@ -0,0 +1,50 @@
+        pha = make_data_path("3c273.pi")
+        ui.load_pha(pha)
+
+    io.backend = orig_backend
```
Technically the test should be something like

```python

pha = mane_data_path('3c273.pi')
ui.load_pha(pha)

orig_backend = io.backend
try:
    import sherpa.astro.io.dummy_backend
    io.backend = sherpa.astro.io.dummy_backend

    with pytest.raises(NotImplementedError,
                       match="No usable I/O backend was imported."):
        ui.load_pha(pha)

finally:
    io.backend = orig_backend

ui.load_pha(pha)
```

just in case there's any problem when switching the backend so we don't ruin other tests, but I'm not sure it's worth it/doesn't cause more problems than is stops

hamogu replied:
It's a test. If it fails something is wrong and the backend can't be switched back, then that needs to be fixed. In principle, I don't think what you suggest is the right idea, but I think the way I've written it is easier. If it works, it's the same. If it fails it needs to be fixed, finally clause or not.

DougBurke commented:
It has long been an annoyance to me that you couldn't import `sherpa.astro.io` if you didn't have a backend, so this is a good thing to fix.

DougBurke commented on sherpa/astro/ui/tests/test_astro_ui_unit.py:
```diff
@@ -640,18 +640,18 @@ def test_save_data_data2d(tmp_path):
         expected = "\n".join([str(zz) for zz in z]) + "\n"
     else:
-        raise RuntimeError(f"UNKNOWN I/O BACKEND: {backend.__name__}")
+        raise RuntimeError(f"UNKNOWN I/O BACKEND: {io.backend.__name__}")
```
I guess this could just be

```
assert False, io.backend.__name__
```

but it's not really worth changing.

hamogu replied:
I'm just following the previous convention here, wither is fine.

DougBurke reviewed (changes requested):
I strongly suggest reverting the sherpa-standalone rc removal

DougBurke commented:
I am also wondering about the removal of the backend option, as we could still take this PR and keep that option, and then have the default setting be a space-separated list of the backends to try in order (so change the default to `pyfits crates`).

This would then make it easier to support your "rewrite pyfits to astropy" PR #1091 as we could have two versions (an astropy one and a pyfits one) with the new development in a new backend called astropy and the old code remaining in pyfits.

[in reality we are unlikely to have any other backend unless an external user wants one, but I am a bit more reluctant to change the status quo here than @hamogu ;-]

DougBurke commented on sherpa/astro/io/__init__.py:
```diff
@@ -17,16 +17,34 @@
+
+When this module is first imported, Sherpa tries to import the
+backends installed with Sherpa in the order listed in the
+``sherpa.rc`` or ``sherpa-standalone.rc`` file. The first module that imports
```
I wonder if it's better to just say `Sherpa configuration file` rather than give the full names.

If you do give the names then we want

```suggestion
``.sherpa.rc`` or ``.sherpa-standalone.rc`` file. The first module that imports
```

DougBurke replied:
Although I guess the file name depends on whether we are reading in the user's version (which has the leading `.`) or the system version (which is dotless). Which makes me think just referring to the configuration file (and possibly mentioning `sherpa.get_config()`) is the better solution.

hamogu replied:
I'm worried that users don't know what the "configuration file" is, in particular since we have two. I thought about "``sherpa.rc`` (if you installed sherpa through CIAO) or ``sherpa-standalone.rc`` (if you installed sherpa as a Python package independent of CIAO)" but found that too verbose. At the very least, when giving both names, users know what name to search for if they don't know the location of this file and they know there is more than one, so if one does not work they will try the other.

DougBurke commented on sherpa/astro/io/__init__.py:
```diff
@@ -56,8 +74,8 @@
-io_opt = config.get('options', 'io_pkg', fallback='pyfits')
-io_opt = str(io_opt).strip().lower()
+io_opt = config.get('options', 'io_pkg', fallback='dummy')
+io_opt = [str(o).strip().lower() + '_backend' for o in io_opt.split()]
```
```suggestion
io_opt = [o.strip().lower() + '_backend' for o in io_opt.split()]
```

I'm pretty sure we don't need the str here. I don't know why it was in the old version (probably defensive programmning).

DougBurke commented on sherpa/astro/io/__init__.py:
```diff
@@ -74,26 +92,24 @@
-    If you are using CIAO, this is most likely an error and you should contact the CIAO helpdesk.
-    If you are using Standalone Sherpa, please install astropy."""
-                      .format(io_opt))
+if backend is None:
```
Is this check needed? You've just set `backend` to `None` which implies this is always going to fire, doesn't it?

DougBurke commented on sherpa/astro/io/dummy_backend.py:
```diff
@@ -0,0 +1,54 @@
+set_image_data = get_table_data
+set_pha_data = get_table_data
+get_column_data = get_table_data
+get_ascii_data = get_table_data
```
So, we do have `sherpa.io.get_ascii_data` which we could use here [modulu potential interface differences], **BUT** I am not going to suggest we do that here - I don't see the point in having a partially usable module here.

DougBurke commented on sherpa/astro/io/tests/test_backend_switch.py:
```diff
@@ -0,0 +1,48 @@
+         ui.load_pha(pha)
+
+    io.backend = orig_backend
+    ui.load_pha(pha)
```
So, normally I would say calling `load_pha` is enough here (ie we can assume it's worked), but given that we are messing around with the backend I would like to see something a bit more than just this. Thoughts are

a) for the first and last `load_pha` calls we do

```python
ui.clean()
ui.load_pha(pha)
assert ui.get_data().name.endswith('3c273.pi')
```

I'm not really convinced it improves things other than being a little-more explicit of the check.

Thoughts?
which just checks we have actually loaded the file (to some very-minor degree).

b) alternatively we make the last `load_pha` call load a different PHA file and check the name has changed (ie it is very similar to a).

DougBurke replied:
Oh, that comment got messed up somewhat...

DougBurke commented:
For reference the summary/details text needs updating to reflect the de-scoping of this PR.

hamogu commented:
Addressed review comments, rebased to pick up recent changes and tests are now passing. I've removed any changes to sherpa.rc that were originally part of this PR and thus it no longer needs review by DS.

DougBurke commented:
Can we squash the commits into one, as I don't think we benefit from having code added then removed?

Marie-Terrell commented on sherpa/astro/io/__init__.py:
```diff
@@ -17,16 +17,34 @@
+backends installed with Sherpa in the order listed in the
+``.sherpa.rc`` or ``.sherpa-standalone.rc`` file. The first module that imports
+successfully is set as the active backend. The following command prints the
+name and the location on disk of that module::
```
```suggestion
name and the location on disk of that module:
```

Marie-Terrell replied:
Thank you for the heads up! I didn't know that.

Marie-Terrell commented on sherpa/astro/io/__init__.py:
```diff
@@ -17,16 +17,34 @@
+successfully is set as the active backend. The following command prints the
+name and the location on disk of that module::
+
+   >>> import sherpa.astro import io
```
```suggestion
   >>> from sherpa.astro import io
```

Marie-Terrell commented:
Other than the two comments, this looks good from the CIAO side for me. One additional request, could you update the rc files to include a sherpa version update at the top (4.14.1)?
Also, not for this PR, but for a future change, should we swap pyfits references to astropy.io.fits? Or, is there a particular reason that this is still referred to as pyfits?

hamogu commented:
> Also, not for this PR, but for a future change, should we swap pyfits references to astropy.io.fits? Or, is there a particular reason that this is still referred to as pyfits?
Just tradition. It used to be called pyfits, so we never changed it. I am working on a new astropy-based backend that will use some more modern features (e.g. use the Table reader for tables instead io.fits) and when I have it, we can call that one "astropy"!

hamogu commented:
Squashed commits.

DougBurke commented:
LGTM - I'll approve it when the tests pass. Thanks for dealing with all the faff above!

DougBurke commented:
Although the test failed it's not related to the code - it's only in the macOS upload, which reports the helpful output below -  so I'm going to approve this.

```
Run codecov/codecov-action@v2
  with:
    files: /home/runner/work/sherpa/sherpa/coverage.xml,/Users/runner/work/sherpa/sherpa/coverage.xml,/home/runner/coverage.xml
    verbose: true
  env:
    sherpa_channel: sherpa
    xspec_channel: xspec/label/test
    miniconda_loc: /Users/runner/work/sherpa/sherpa/miniconda
    CONDA_BUILD_SYSROOT: /Users/runner/work/sherpa/sherpa/10.9SDK/MacOSX10.9.sdk
==> macos OS detected
https://uploader.codecov.io/latest/macos/codecov.SHA256SUM
Error: Codecov: Error validating uploader: Misformed armored text
```

DougBurke reviewed (approved):
:shipit:

DougBurke commented:
Just to note that there's something to think about in #1354 - but this is not because of this PR; we'd have had something to do there even without this PR (i.e. it's not a reason not to merge this).

DougBurke commented:
:1st_place_medal:
